src/data.ts

```diff
@@ -1,54 +1,72 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
 // import { relayPool } from 'nostr-tools';
 // import { validateEvent } from 'nostr-tools';
-import { relayPool } from './lib/nostr-tools';
+import { relayPool, type RelayPool } from './lib/nostr-tools';
 import { validateEvent } from './lib/nostr-tools';
 import { db } from './db';
 import type { IProfile, IEvent } from './db';
+import { snooze, yieldMicrotask } from './utils/sync';
+import { filterMap, filterTF } from './utils/array';
+import { deepCloneObj } from './utils/clone';
 
-export class Data {
-	private static _instance: Data = new this();
-	public pool;
-	private allEventIds: Set<string>
+const itemKinds = [0, 1, 3, 4, 5, 7];
 
+export class $Data {
+	public pool!: RelayPool;
+	private allEventIds!: WeakSet<IEvent>
 	// event buffer to batch verify and insert
-	public events: object[] = [];
-
-	private constructor() {}
+	public events: (IProfile | IEvent)[] = [];
+	constructor() {
+		//!Empty
+	}
 
 	public async start() {
-		this.allEventIds = new Set((await db.events.toArray()).map(it => it.id));
-		this.storeEventsLoop();
+		this.allEventIds = new WeakSet((await db.events.toArray()).map(it => it));
+		// Use the `snooze` function to defer intensive code execution since
+		// each of these calls happen synchronously
+		this.storeEventsLoop()
+		await snooze(50)
 		this.loadAndWatchProfiles();
-		db.updateProfileFromMeta();
+		await snooze(50)
+		//? Are we sure we need to call this without a parameter?
+		//@ts-expect-error Original implementation
+		await db.updateProfileFromMeta();
+		await snooze(50)
 	}
 
 	// check for work constantly
-	private async storeEventsLoop() {
-		const snooze = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
-		while (true) {
-			Data.instance.events = Data.instance.events.filter(
+	private async storeEventsLoop(): Promise<never> {
+		let delay = 500;
+		let promiseQueue = [];
+		while (true){
+			delay = 500;
+			this.events = (Data.events as IEvent[]).filter(
 				(value, index, self) => index === self.findIndex((t) => t.id === value.id)
 			);
 			const t = Date.now();
-			const e = Data.instance.events.splice(-500);
+			const e = Data.events.splice(-500);
 			if (e.length > 0) {
+				delay = 250;
 				e.forEach((event) => {
-					event.tags = event.tags?.map((it) => it.join('»')) || [];
+					(event as IEvent).tags = (event as IEvent).tags?.map((it) => (it as unknown as string[]).join('»')) || [];
 				});
-				await db.events.bulkPut(e);
+				promiseQueue.push(db.events.bulkPut(e as IEvent[]));
 				const dt = Date.now() - t;
 				console.log(`It took ${dt}ms to store ${e.length} events.`);
```

Comment:
Measuring time makes no sense any more after making the bulkPut call non-blocking.